Fixed `wp-env start` On Windows

ObliviousHarmony opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

This pull request assigns an arbitrary `uid` and `gid` of `1000` when running `wp-env` from Windows natively (without using WSL).

Closes #50814.

## Why?

When we fixed the permission-related problems we introduced an incompatibility when running `wp-env` using Windows natively. While it _does_ work with WSL, when used natively, the `uid` and `gid` are `-1`. We handled this case by assigning it a `uid` and `gid` of `0`, but, this introduced other problems with trying to run everything as `root`.

## How?

We just assign an arbitrary `uid` and `gid`. This _should_ work fine because Windows is already performing magic for us with filesystem permissions to get everything in working order in the first place. It doesn't really matter to Windows what the user is because the Docker container has full permissions.

## Testing Instructions

1. Install `node` on Windows without using WSL. Meaning you should be able to run it from PowerShell or CMD.
2. Run `npx wp-env start` from Windows without using WSL. It should boot correctly without any problems.
3. Try to install a plugin. We expect this to work without problem.

ObliviousHarmony committed:
Fixed User Switch Bug

Since it's possible for the user creation to fail, it's possible
that there will be no user by the given name to switch to.
To prevent any errors we should switch to the UID:GID so
that if the username was not created we don't error out.

ObliviousHarmony committed:
Fixed Windows `uid` and `gid`

noahtallen reviewed (approved):
The changes make sense to me; I'll let you and @t-hamano test it on Windows!

ObliviousHarmony committed:
Fixed `wp-env run` Windows Path Resolution

ObliviousHarmony committed:
Fixed Windows `composer` Paths

ObliviousHarmony committed:
Fixed Broken `sudoers` On Windows

ObliviousHarmony committed:
Use `uid` In Sudoers Files

ObliviousHarmony committed:
Set FS_METHOD constant to 'direct'

ObliviousHarmony commented:
Thanks @noahtallen,

In the process of going through this pull request on Windows (natively) I managed to find a couple more issues with Windows that I resolved.

- `wp-env run`'s relative path resolution was breaking on Windows because paths are made absolute from the `C:\` drive. I switched to using POSIX paths so that it has the correct result.
- Using `$HOST_USERNAME` in the Dockerfile causes a lot of problems when the user doesn't exist. I've changed the cases where it causes problems to use the `uid` instead so that it always works.
- Plugin installation on Windows wasn't working because WordPress felt like it didn't have filesystem permissions but _does_. Setting `FS_METHOD` to `'direct'` fixes this, and in the case of our `wp-env` environments, it shouldn't cause any trouble.

I've thoroughly tested these changes on both Windows and macOS to confirm that they don't break anything. It looks like Windows is working completely.

@t-hamano I'm going to avoid merging this _just yet_, but, I'd like to get it in before tomorrow's automatic package release that way Windows will be functional again.

ObliviousHarmony committed:
Changelog

t-hamano reviewed (approved):
Thank you for the quick response! Everything worked as expected.

I tested on a Windows _host_ OS using the following commands:

```
git fetch origin fix/wp-env-windows-start
git checkout fix/wp-env-windows-start 
npm run wp-env destroy
npm ci
npm run build
npm run wp-env start
```
- ✅ Started up with no problem
- ✅ Install and remove themes
- ✅ Install and remove plugins
- ✅ Uploading media
- ✅ Updating code using the theme editor
- ✅ Switching languages (i.e., downloading language files)

ObliviousHarmony commented:
Thanks for the thorough review @t-hamano!
